### ISSUE TITLE
What is the difference between two tensorflow inception models 'inception_v3_2016_08_28.tar.gz' and 'classify_image_graph_def.pb'

### ISSUE THREAD

**Issue by nirmaljith** [tensorflow/models#1314]
**Tensorflow version 1.0**
**Python 2.7**
My dataset has got two classes. For transfer learning i tried two inceptionV3 models.
  1. **inception_v3_2016_08_28.tar.gz**
  2. **classify_image_graph_def.pb** - comes along with tensorflow image_retraining code.

But the results i am getting are totally different. Where the second models performs way better than the first. Is it expected ? First model gives accuracy of 57% where the second model gives 80% accuracy.

The first model is checkpoint file. For transfer learning, i have converted the checkpoint file to protobuf file. Then the python code retrain.py which comes along with tensorflow is used to do the retraining. Following code is used to convert checkpoint file to a protobuf file **inceptionv3.pb**.

	checkpoint_file = '../check_points/inception_v3.ckpt'
	decode_jpeg_data = tf.placeholder(tf.string)
	decode_jpeg = tf.image.decode_jpeg(decode_jpeg_data, channels=3, dct_method="INTEGER_ACCURATE")
	if decode_jpeg.dtype != tf.float32:
	  decode_jpeg = tf.image.convert_image_dtype(decode_jpeg, dtype=tf.float32)
	image_ = tf.expand_dims(decode_jpeg, 0)
	image = tf.image.resize_bicubic(image_, [299, 299], align_corners=True)
	scaled_input_tensor = tf.scalar_mul((1.0/255), image)
	scaled_input_tensor = tf.subtract(scaled_input_tensor, 0.5)
	scaled_input_tensor = tf.multiply(scaled_input_tensor, 2.0)

	# loading the inception graph
	arg_scope = inception_v3_arg_scope()
	with slim.arg_scope(arg_scope):
	  logits, end_points = inception_v3(inputs=scaled_input_tensor, is_training=False, num_classes=1001)
	saver = tf.train.Saver()
	sess = tf.Session()
	saver.restore(sess, checkpoint_file)
	with gfile.FastGFile('./models/inceptionv3.pb', 'wb') as f:
	    f.write(output_graph_def.SerializeToString())

Am I missing something while converting checkpoint files to graph ? 
Can i use the retrain.py code to do the tranfer learning of the graph obtained using above code?
 
Also asked the same question in stackoverflow as well, but no luck 
http://stackoverflow.com/questions/43246387/what-is-the-difference-between-two-tensorflow-inception-models-inception-v3-201

**Comment by nealwu** [comment]
Re-assigning to the slim/inception maintainers: @sguada @nathansilberman

**Comment by nirmaljith** [comment]
@sguada any update on this issue

**Comment by freud14** [comment]
Hi, I have the exact same problem. I tried with other models (resnet, inception-resnet) but are all ~50% of accuracy whereas I get ~90% with classify_image_graph_def.pb.

**Comment by freud14** [comment]
Any news on that?

**Comment by nirmaljith** [comment]
@nealwu @sguada @nathansilberman , guys any update on this. It will be of great help at-least if you mention the difference between the two models.

**Comment by korrawat** [comment]
@nirmaljith @freud14 What was the code you used for retraining? Do you mean this [`retrain.py`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/image_retraining/retrain.py#L100-L114)? If so, have you tried changing model-specific parameters such as `BOTTLENECK_TENSOR_NAME`?

See also: I just filed an [issue](https://github.com/tensorflow/tensorflow/issues/11181) on the main tensorflow repo about the two models. (As you can see from that post, I'm still confused with the differences, but I hope it helps.)

**Comment by freud14** [comment]
>  Do you mean this retrain.py? If so, have you tried changing model-specific parameters such as BOTTLENECK_TENSOR_NAME?

Yes, this retrain. Yes, I've changed model-specific parameters.

**Comment by chenyuZha** [comment]
I have the same issue about this 2 models. When I trained my own datasets with **retrain.py**(version 2015.12 of inception v3), my final accuracy was about **88%**, then with the **checkpoint inception v3** (version 2016.8), accuracy went down to **70%**.
 I noticed that  in the checkpoint version, two scopes are excluded : **InceptionV3/AuxLogits, InceptionV3/Logits**, then we retrained these two scopes. Though in the **retrain.py**, it seems like 
to just retrain the **bottleneck(ArgPool) layer**...

**Comment by freud14** [comment]
> I noticed that in the checkpoint version, two scopes are excluded : InceptionV3/AuxLogits, InceptionV3/Logits, then we retrained these two scopes.

Do you mean that you got 70% of accuracy when retraining these two scopes or that you got the initial 88% as with the version 2015.12 of inception v3?

**Comment by oneTimePad** [comment]
I have been experiencing strange issues with `inception_v3_2016_08_28.tar.gz` as well. I was performing simple transfer learning on a flowers dataset to train a new output layer. I was able to achieve 99% accuracy on the training set. However, when I would restore my model and it was evaluated on the random batches of the training set, the accuracy dropped to about 70%. What I noticed was that when I set `is_training` back to `True` when I was performing inference, the accuracy jumped back up to 99% for training and around 95% for the test set. I don't quite understand how enabling dropout to drop units during inference would increase the accuracy...

**Comment by JasonHJG** [comment]
How do you find the three tensors defined for version 2015.12 in version 2016? I have difficulty in finding and loading the correct tensors.

**Comment by korrawat** [comment]
@JasonHJG Try the following:
- [`import_pb_to_tensorboard.py`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/import_pb_to_tensorboard.py)
- [`summarize_graph`](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/graph_transforms#inspecting-graphs) (Another example [here](https://github.com/tensorflow/models/blob/master/slim/export_inference_graph.py#L34))

**Comment by chenyuZha** [comment]
@freud14 
The 70% accuracy is obtained with the checkpoint version of InceptionV3, I didn't change the option `--checkpoint_exclude_scopes` of file `train_image_classifier.py `, so by default, two scopes are excluded:  **InceptionV3/AuxLogits, InceptionV3/Logits** and we do fine-tuning with 2 scopes. While the 88% is obtained with `retrain.py`, where we calculate the **bottlenecks_values**  for new datasets, then these values are considered like the feature extractors for the classification layer.

**Comment by freud14** [comment]
@chenyuZha 
It seems normal. In the case where you are using `train_image_classifier.py`, it retrains the whole model so it will take a longer time to get the same accuracy as `retrain.py`.

**Comment by chenyuZha** [comment]
@freud14 
I'm not sure if `retrain.py` will work with 2016 checkpoint even though I convert it into a .pb file, since the checkpoint version has modified the name of nodes. So, we may not find the same `input_tensor` and `final_result` , even the layer which store the bottleneck values.
Also, the 2016 checkpoint took longer time to train than version 2015 because the weights are also recalculated ( not from beginning but just 2 scopes **InceptionV3/AuxLogits,InceptionV3/Logits** in my opinion ), so normally these weights should adapt well to your own datasets, and the accuracy should be higher than version 2015, that's why I'm confused.

**Comment by freud14** [comment]
> I'm not sure if retrain.py will work with 2016 checkpoint even though I convert it into a .pb file, since the checkpoint version has modified the name of nodes. So, we may not find the same input_tensor and final_result , even the layer which store the bottleneck values.

Yeah, you have to change the values of the constants for it to work. [Here](https://gist.github.com/freud14/fc364781c2b9f2e48d52a7b772ab5ac4) are the scripts I used in my case.

>  Also, the 2016 checkpoint took longer time to train than version 2015 because the weights are also recalculated ( not from beginning but just 2 scopes InceptionV3/AuxLogits,InceptionV3/Logits in my opinion ), so normally these weights should adapt well to your own datasets, and the accuracy should be higher than version 2015, that's why I'm confused.

The script `train_image_classifier.py` takes your checkpoint and initialize your model with it. It removes the last layers (InceptionV3/AuxLogits,InceptionV3/Logits; dunno what really is InceptionV3/AuxLogits) and replace them with new layers with the number of classes you specified. These new layers are initialized randomly as in `retrain.py`. Then, all the weights of your new model are retrain. So, instead of just retraining the last layers,  `train_image_classifier.py` retrains the whole model. To get a better accuracy with this model, you would need to retrain it for much much longer than with `retrain.py` since you have to forward-backward propagate through the whole network for each image.

**Comment by freud14** [comment]
Anyway, I suspect that inception-v3-2016 was not trained as much as the 2015 model.

**Comment by chenyuZha** [comment]
@freud14 Thanks for your file and explication. So I see that you change the `BOTTLENECK_TENSOR_NAME` to the 2016 checkpoint version. So in this case the bottlenecks values are calculated just once and then you can use it to train many times as you can to save time. I wonder that the bottlenecks values are not the same with 2015 version for the same datasets..? So how much accuracy do you get ?

**Comment by freud14** [comment]
Alright, after reading @oneTimePad's comment (which I actually didn't understand completely), I decided to check whether the is_training boolean was set to true or not and it was set to true. So I put it to false and did the training and testing again and now I get similar results as with the 2015 model. It must have been something to do with the moving average but I don't understand really what.

Anyway, I will try the other models (resnet, resnet-inception and other versions of inception) soon to see if it was really only that (at least for me).

**Comment by XYAskWhy** [comment]
@freud14 Sorry to bother, but it seems that your "save_inception_v3.py" doesn't wokr for me. Any help please? And your "retrain_inception_v3.py" was based on the older version of "retrain.py", what about the newer one which doesn't include the puzzling "JPEG_DATA_TENSOR_NAME"?

**Comment by chenyuZha** [comment]
@freud14 
I check also `is_training` option, so for the inception model, the images are processed in the file `inception_preprocessing.py`, inside there are two functions `preprocess_for_eval` and  `preprocess_for_train`. If `is_training=False`, all of input images will go through the function `preprocess_for_eval`.

The difference between these two functions are: `preprocess_for_train` will distort input image.

**Comment by freud14** [comment]
@XYAskWhy I don't know, I use an older version of Tensorflow so that must be why. You can probably adapt the scripts to the new version. For "save_inception_v3.py", I notice that there is a [new script](https://github.com/tensorflow/models/blob/master/slim/export_inference_graph.py) in the models repo that looks like it should do the same thing.

@chenyuZha I do not use this preprocessing script. The `is_training` variable is also used to enable/disable dropout and also to set how batch normalization is done.

**Comment by oneTimePad** [comment]
@nirmaljith have you tried setting `is_training` to True  when you are testing the accuracy and not training? I know this doesn't make sense but it worked for me. No idea why.

**Comment by oneTimePad** [comment]
[Here](https://stackoverflow.com/questions/45134294/reporting-accuracy-and-loss-issues-with-monitoredtrainingsession?noredirect=1#comment77365725_45134294) is the link to the Stackoverflow post with a comment that helped resolved my issue,  my interpretation of it is described below. This might explain why `is_training=False` caused the accuracy of the 2016 model for @freud14  to increase significantly. It might or might not be related to the issues with the 2016 model not performing well.  If `is_training` was set to True when both training and testing and you still received low accuracy, then this probably doesn't explain the issue. For that case, I have no idea. However, I was able to achieve high accuracy on both the training and test set for  transfer learning using the 2016 model utilizing the method below. I hope this helps someone out.

When `is_training=True` batch normalization is computing the sample mean and sample variance of a batch. A parameter update is then done under the assumption that the output of a convolutional layer with batch norm is normalized using that layer's input's sample mean and variance. These sample means and variances are supposed to contribute the long running means and variances which will be used as the population statistics at test time. However, if the `UPDATE_OPS` are not ran during training, the new long running statistics taking your new data into account are never actually written to variables. 

 I am not sure if anyone tried this, but I bet if after each mini batch update you tested the accuracy of the model, with `is_training=True`, on that same mini-batch, you would see the accuracy significantly increase over time.  This is exactly what happened to me and why I got 99% accuracy during training, but not after training completed. Then if you went to restore the model, set `is_training=False`, and tested the accuracy on mini batches of  the training set, you would receive very low accuracy. This is because, based on the Stackoverflow comment talking about UPDATE_OPS, the model is now utilizing the long running average and variance that were computed for ImageNet, because there were no update ops during training. Thus the weights and biases learned during the training of your new dataset were updated based on the fact that the sample mean and variance of all inputs to batch norm were utilized to normalize that input; this I believe is the "transient component" the comment in the Stackoverflow post is referring to. Thus the input to layers during training time is very different from the inputs during test time.

I believe the reason the model now performs better with `is_training=False` is because the long running average and variance obtained from when the model was originally trained on ImageNet are now being used both for training and testing. My model also performed better when I did that both during training and testing.

I would suggest training the model again with `is_training=True`, and with the following modification:
      ```
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
      with tf.control_dependencies([update_ops]):
                  optimizer.minimize(loss,var_list=(variables to train))  or whatever you used
```
Note: When you attempt to get the UPDATE_OPS via get_collection, `is_training` must be set to True, or else the Tensorflow  Inception implementation won't add the update ops to the graph collections.

When I  then restored the model, after training it with `is_training=True`, and check the accuracy on the data with `is_training` now False, the accuracy matched what I was getting when computing it with `is_training=True`. 

With this modification I received high accuracy on both the training and test set with the 2016 checkpoint.

**Comment by XYAskWhy** [comment]
@freud14 Finally I got it right with the 2016 inception_v3 model retraining. Many thanks! By the way, what's your result with all the other models?

**Comment by Steven-N-Hart** [comment]
@freud14  Just to be explicit here, are you proposing the following at line 486 in [train_image_classifier.py](https://github.com/tensorflow/models/blob/master/slim/train_image_classifier.py#L486)?
```
# Gather update_ops from the first clone. These contain, for example,
# the updates for the batch_norm variables created by network_fn.
# Add summaries for losses.
for loss in tf.get_collection(tf.GraphKeys.LOSSES, first_clone_scope):
      summaries.add(tf.summary.scalar('losses/%s' % loss.op.name, loss))
      with tf.control_dependencies([update_ops]):
           optimizer.minimize(loss,var_list=(slim.get_model_variables()))
                  optimizer.minimize(loss,var_list=(variables to train))
```

**Comment by freud14** [comment]
@XYAskWhy I have not tested yet the other models. Will come back when when I have done it.

@oneTimePad That makes sense for both problems. What do you use for training? Because it does not seem possible to modify the value of `is_training` with the `retrain.py` script.

**Comment by chenyuZha** [comment]
@freud14 @oneTimePad  I also have the same question. I found `is_training` both in the `train_image_classifier.py` and `eval_image_classifier.py`.

In the same time, I tried the model InceptionV4, for this model, the accuracy didn't change much. (90% vs 89.1%). So I suppose InceptionV4 is ok , no need to change `is_training` boolean.
Then, I tried model Inception ResnetV2,  and accuracy is chute..( 70% vs 89.1%). I'll try to change the `is_training` to see if there will be a progress.

**Comment by freud14** [comment]
@chenyuZha If the objects to detect in your dataset is relatively small in your images, it might be just because the size of the input image in Resnet (224x224) is smaller than in Inception (299x299). However, since Resnet does not contain any fully connected layer, you can try to change the input size to see if you get better results.

**Comment by chenyuZha** [comment]
@freud14  In fact the model I used is Inception Resnet v2, not resnet v2, so the first one is the one which has the highest accuracy so far (95.3% for top-5 in ImageNet), so the input image is resized a 299x299.
Anyway, I will check `is_training` boolean to see if I get a better accuracy.

**Comment by karmel** [comment]
This question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. There is also a larger community that reads questions there. Thanks!

**Comment by rewreu** [comment]
Is that possible there is dropout path used for the inference makes the accuracy goes down from the checkpoint?

**Comment by JoeyChuuichi** [comment]
I have experienced the same issue. The accuary by training from **classify_image_graph_def.pb** is **95%** while the accuary from **inception_v3_2016_08_28.tar.gz** is about 30%. That is truly abnormal.

**Comment by JoeyChuuichi** [comment]
I have experienced the same issue. The accuary by training from classify_image_graph_def.pb is 95% while the accuary from inception_v3_2016_08_28.tar.gz is about **30%**. That is truly abnormal.

**Comment by JoeyChuuichi** [comment]
I have experienced the same issue. The accuary by training from classify_image_graph_def.pb is **95%** while the accuary from inception_v3_2016_08_28.tar.gz is about **30%**. That is truly abnormal.
